File: Game.java
package card51;

import java.util.*;
import static card51.GroupValidator.AceMode;

public class Game {
    private Deck deck = new Deck();
    private List<Player> players = new ArrayList<>();
    private Scanner input = new Scanner(System.in);

    public Game(int numPlayers) {
        for (int i = 0; i < numPlayers; i++)
            players.add(new Player());
        deal();
    }

    private void deal() {
        for (Player p : players)
            for (int j = 0; j < 14; j++)
                p.addCard(deck.draw());
    }

    public void start() {
        int turn = 0;
        while (true) {
            Player p = players.get(turn);
            System.out.printf("\nPlayer %d's turn%n", turn + 1);
            p.showHand();

            // DRAW PHASE
            System.out.println("Draw from (1) Deck or (2) Discard?");
            int choice = input.nextInt();
            Card drawn;
            if (choice == 2 && p.hasOpened() && !deck.isDiscardEmpty())
                drawn = deck.pullDiscard();
            else {
                drawn = deck.draw();
                if (choice == 2) {
                    String msg = !p.hasOpened()
                            ? "You have not yet opened, drawing from the deck."
                            : "Discard pile is empty, drawing from the deck.";
                    System.out.println(msg);
                }
            }
            p.addCard(drawn);
            System.out.println("Drew: " + drawn);

            // OPENING LOGIC
            if (!p.hasOpened()) {
                List<List<Card>> candidates = new ArrayList<>();
                candidates.addAll(GameUtils.findSets(p.getHand()));
                candidates.addAll(GameUtils.findRuns(p.getHand(), AceMode.LOW));
                candidates.addAll(GameUtils.findRuns(p.getHand(), AceMode.HIGH));

                List<List<Card>> openables = new ArrayList<>();
                for (List<Card> g : candidates) {
                    if (g.contains(drawn) && GroupValidator.getGroupPoints(g) >= 51)
                        openables.add(g);
                }
                if (!openables.isEmpty()) {
                    System.out.println("You can open! Choose a group:");
                    for (int i = 0; i < openables.size(); i++)
                        System.out.printf("%2d: %s (points %d)%n", i + 1, openables.get(i),
                                GroupValidator.getGroupPoints(openables.get(i)));
                    int sel = input.nextInt() - 1;
                    List<Card> chosen = openables.get(sel);
                    if (p.getHand().size() - chosen.size() < 1) {
                        System.out.println("Must keep one card to discard. Opening canceled.");
                    } else {
                        for (Card c : chosen)
                            p.removeCard(c);
                        p.addGroup(chosen);
                        p.open();
                        System.out.println("Opened with: " + chosen);
                    }
                }
            } else {
                // POST-OPEN: additional melds
                System.out.println("Lay a new meld? (1) Yes (2) No");
                if (input.nextInt() == 1) {
                    List<List<Card>> all = new ArrayList<>();
                    all.addAll(GameUtils.findSets(p.getHand()));
                    all.addAll(GameUtils.findRuns(p.getHand(), AceMode.LOW));
                    all.addAll(GameUtils.findRuns(p.getHand(), AceMode.HIGH));
                    for (int i = 0; i < all.size(); i++)
                        System.out.printf("%2d: %s (points %d)%n", i + 1, all.get(i),
                                GroupValidator.getGroupPoints(all.get(i)));
                    int sel = input.nextInt() - 1;
                    List<Card> chosen = all.get(sel);
                    if (p.getHand().size() - chosen.size() < 1) {
                        System.out.println("Must leave one card to discard.");
                    } else {
                        for (Card c : chosen)
                            p.removeCard(c);
                        p.addGroup(chosen);
                        System.out.println("Laid meld: " + chosen);
                    }
                }
            }

            p.showHand();
            // DISCARD PHASE
            System.out.println("Select card to discard:");
            int d = input.nextInt() - 1;
            Card disc = p.getHand().get(d);
            p.removeCard(disc);
            deck.discard(disc);
            System.out.println("Discarded: " + disc);

            // WIN CHECK
            if (p.getHand().isEmpty()) {
                System.out.printf("Player %d wins!%n", turn + 1);
                break;
            }
            turn = (turn + 1) % players.size();
        }
    }

    public static void main(String[] args) {
        System.out.println("Welcome to Card 51!");
        new Game(2).start();
    }
}

File: game.ts
// src/engine/game.ts
//
// This is a direct TypeScript port of your Java Game loop,
// but with ONE critical change: no Scanner/console input.
// Instead, the Game asks a "DecisionProvider" for choices.
// Your web UI will implement DecisionProvider later.

import { Deck } from "./deck";
import { Card } from "./card";
import { Player } from "./player";
import { GroupValidator, AceMode } from "./groupValidator";
import { GameUtils } from "./gameUtils";

export interface DecisionProvider {
  /** Return 1 for Deck, 2 for Discard. */
  chooseDrawSource(ctx: { playerIndex: number; canDrawDiscard: boolean; discardTop?: Card }): 1 | 2;

  /**
   * Return index (0-based) into openables, or null to skip opening.
   * UI should present groups and points to the user.
   */
  chooseOpenGroup(
    ctx: { playerIndex: number; drawn: Card },
    openables: ReadonlyArray<ReadonlyArray<Card>>,
    points: ReadonlyArray<number>
  ): number | null;

  /** After opening, decide whether to lay a new meld. */
  chooseLayNewMeld(ctx: { playerIndex: number }): boolean;

  /**
   * Choose a meld from all candidates (0-based), or null to skip.
   * Only called if chooseLayNewMeld returned true.
   */
  choosePostOpenMeld(
    ctx: { playerIndex: number },
    candidates: ReadonlyArray<ReadonlyArray<Card>>,
    points: ReadonlyArray<number>
  ): number | null;

  /** Choose a discard card index (0-based) from current hand. */
  chooseDiscardIndex(ctx: { playerIndex: number }, hand: ReadonlyArray<Card>): number;
}

export class Game {
  private deck: Deck = new Deck();
  private players: Player[] = [];
  private turn: number = 0;

  constructor(numPlayers: number) {
    for (let i = 0; i < numPlayers; i++) this.players.push(new Player());
    this.deal();
  }

  private deal(): void {
    for (const p of this.players) {
      for (let j = 0; j < 14; j++) p.addCard(this.deck.draw());
    }
  }

  /**
   * Runs until someone wins.
   * The UI supplies choices through DecisionProvider.
   *
   * Returns the winning player index (0-based).
   */
  public run(decider: DecisionProvider): number {
    while (true) {
      const p = this.players[this.turn];

      // DRAW PHASE
      const canDrawDiscard = p.hasOpened() && !this.deck.isDiscardEmpty();
      const drawChoice = decider.chooseDrawSource({
        playerIndex: this.turn,
        canDrawDiscard,
        discardTop: this.deck.peekDiscard(),
      });

      let drawn: Card;

      if (drawChoice === 2 && canDrawDiscard) {
        drawn = this.deck.pullDiscard();
      } else {
        drawn = this.deck.draw();
        // In Java you printed a message when they tried discard but couldn't.
        // UI can show that based on canDrawDiscard.
      }

      p.addCard(drawn);

      // OPENING LOGIC (must include the drawn card, and group points >= 51)
      if (!p.hasOpened()) {
        const candidates: Card[][] = [
          ...GameUtils.findSets(p.getHand()),
          ...GameUtils.findRuns(p.getHand(), AceMode.LOW),
          ...GameUtils.findRuns(p.getHand(), AceMode.HIGH),
        ];

        const openables: Card[][] = [];
        const pts: number[] = [];

        for (const g of candidates) {
          if (g.includes(drawn)) {
            const gp = GroupValidator.getGroupPoints(g);
            if (gp >= 51) {
              openables.push(g);
              pts.push(gp);
            }
          }
        }

        if (openables.length > 0) {
          const sel = decider.chooseOpenGroup({ playerIndex: this.turn, drawn }, openables, pts);

          if (sel !== null && sel >= 0 && sel < openables.length) {
            const chosen = openables[sel];

            // must keep at least one card to discard
            if (p.getHand().length - chosen.length < 1) {
              // opening cancelled (UI can display)
            } else {
              for (const c of chosen) p.removeCard(c);
              p.addGroup(chosen);
              p.open();
            }
          }
        }
      } else {
        // POST-OPEN: additional melds
        const wantsMeld = decider.chooseLayNewMeld({ playerIndex: this.turn });

        if (wantsMeld) {
          const all: Card[][] = [
            ...GameUtils.findSets(p.getHand()),
            ...GameUtils.findRuns(p.getHand(), AceMode.LOW),
            ...GameUtils.findRuns(p.getHand(), AceMode.HIGH),
          ];

          const pts = all.map((g) => GroupValidator.getGroupPoints(g));
          const sel = decider.choosePostOpenMeld({ playerIndex: this.turn }, all, pts);

          if (sel !== null && sel >= 0 && sel < all.length) {
            const chosen = all[sel];
            if (p.getHand().length - chosen.length < 1) {
              // must leave one to discard (UI can display)
            } else {
              for (const c of chosen) p.removeCard(c);
              p.addGroup(chosen);
            }
          }
        }
      }

      // DISCARD PHASE
      const hand = p.getHand();
      const discardIndex = decider.chooseDiscardIndex({ playerIndex: this.turn }, hand);

      if (discardIndex < 0 || discardIndex >= hand.length) {
        throw new Error(`Invalid discard index ${discardIndex} for hand size ${hand.length}.`);
      }

      const disc = hand[discardIndex];
      p.removeCard(disc);
      this.deck.discard(disc);

      // WIN CHECK
      if (p.getHand().length === 0) {
        return this.turn;
      }

      this.turn = (this.turn + 1) % this.players.length;
    }
  }

  // Useful for UI
  public getPlayers(): ReadonlyArray<Player> {
    return this.players;
  }

  public getCurrentPlayerIndex(): number {
    return this.turn;
  }

  public getDeck(): Deck {
    return this.deck;
  }
}
